common/minicurl.cc

```diff
@@ -170,3 +202,156 @@ bool MiniCurl::postURL(const std::string& url,
+#endif
+  curl_multi_setopt(d_mcurl, CURLMOPT_MAX_HOST_CONNECTIONS, d_ccs.size());
+  d_current = d_ccs.begin();
+}
+
+bool MiniCurlMulti::addPost(unsigned int id, const std::string& url,
+                            const std::string& post_body,
+                            const MiniCurlHeaders& headers)
+{
+  if (d_mcurl) {
+    // return false if we don't have any spare connections
+    if (d_current == d_ccs.end())
+      return false;
+    d_current->setPostData(url, post_body, headers);
+    d_current->setID(id);
+    curl_multi_add_handle(d_mcurl, d_current->getCurlHandle());
+    ++d_current;
+    return true;
+  }
+  return false;
+}
+
+const std::vector<mcmPostReturn> MiniCurlMulti::runPost()
+{
+  std::vector<mcmPostReturn> post_ret;
+  if (d_mcurl) {
+    int still_running;
+    int without_fds = 0;
+    do {
+      CURLMcode mc;
+      int numfds;
+      mc = curl_multi_perform(d_mcurl, &still_running);
+      if (mc == CURLM_OK) {
+        auto start_time = std::chrono::steady_clock::now();
+        
+        mc = curl_multi_wait(d_mcurl, NULL, 0, 1000, &numfds);
+
+        auto wait_time_ms = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start_time);
+        
+        if (mc == CURLM_OK) {
+          struct CURLMsg *m;
+          do {
+            int msgq = 0;
+            m = curl_multi_info_read(d_mcurl, &msgq);
+            if (m && (m->msg == CURLMSG_DONE)) {
+              CURL *c = m->easy_handle;
+              auto minic = findMiniCurl(c);
+              if (minic == d_ccs.end()) {
+                throw WforceException("Cannot find curl handle in MiniCurlMulti - something is very wrong");
+              }
+              struct mcmPostReturn mpr;
+              if (m->data.result != CURLE_OK) {
+                mpr.ret = false;
+                std::string error_msg = minic->getErrorMsg();
+                if (error_msg.length())
+                  mpr.error_msg = error_msg;
+                else
+                  mpr.error_msg = std::string(curl_easy_strerror(m->data.result));
+              }
+              else {
+                long response_code;
+                curl_easy_getinfo(c, CURLINFO_RESPONSE_CODE, &response_code);
+                if ((response_code < 200) || (response_code > 299)) {
+                  mpr.error_msg = std::string("Received non-2XX response from webserver: ") + std::to_string(response_code);
+                  mpr.ret = false;
+                }
+                else {
+                  mpr.ret = true;
+                  mpr.result_data = minic->getPostResult();
+                }
+              }
+              mpr.id = minic->getID();
+              post_ret.emplace_back(std::move(mpr));
+            }
+          } while (m);
+          if (!numfds) {
+            // Avoid busy looping when there is nothing to do
+            // Inspired by similar code in curl_easy_perform
+            if (wait_time_ms.count() <= 10) {
```

Comment:
this code seems somehow overcly complicated. at least it could use some explanation.